docs/proposals/horizontal-pod-autoscaler.md

```diff
@@ -0,0 +1,272 @@
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/horizontal-pod-autoscaler.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Horizontal Pod Autoscaling
+
+**Author**: Jerzy Szczepkowski (@jszczepkowski)
+
+## Preface
+
+This document briefly describes the design of the horizontal autoscaler for pods.
+The autoscaler (implemented as a kubernetes control loop) will be responsible for automatically
+choosing and setting the number of pods of a given type that run in a kubernetes cluster.
+
+This proposal supersedes [autoscaling.md](http://releases.k8s.io/release-1.0/docs/proposals/autoscaling.md).
+
+## Overview
+
+The usage of a serving application usually vary over time: sometimes the demand for the application rises,
+and sometimes it drops.
+In the version 1.0, a user can only manually set the number of serving pods.
+Our aim is to provide a mechanism for the automatic adjustment of the number of pods basing on usage statistics.
+
+## Scale Subresource
+
+We are going to introduce Scale subresource and implement horizontal autoscaling of pods on a base of it.
+Scale subresource will be supported for replication controllers and deployments.
+HorizontalPodAutoscaler object will be bound with exactly one Scale subresource and will be
+autoscaling associated replication controller/deployment through it.
+
+Scale subresource will be present for replication controller or deployment under the following paths:
+
+```api/vX/replicationcontrollers/myrc/scale```
+
+```api/vX/deployments/mydeployment/scale```
+
+It will have the following structure:
+
+```go
+// Scale subresource, applicable to ReplicationControllers and (in future) Deployment.
+type Scale struct {
+	api.TypeMeta
+	api.ObjectMeta
+
+	// Spec defines the behavior of the scale.
+	Spec ScaleSpec
+
+	// Status represents the current status of the scale.
+	Status ScaleStatus
+}
+
+// ScaleSpec describes the attributes a Scale subresource
+type ScaleSpec struct {
+	// Replicas is the number of desired replicas.
+	Replicas int
+}
+
+// ScaleStatus represents the current status of a Scale subresource.
+type ScaleStatus struct {
+	// Replicas is the number of actual replicas.
+	Replicas int
+
+	// Selector is a label query over pods that should match the replicas count.
+	Selector map[string]string
```

Comment:
On Aug 20, 2015, at 11:58 AM, Jerzy Szczepkowski <notifications@github.com>
wrote:

In docs/proposals/horizontal-pod-autoscaler.md
<https://github.com/kubernetes/kubernetes/pull/12859#discussion_r37546968>:

> +	Status ScaleStatus
> +}
> +
> +// ScaleSpec describes the attributes a Scale subresource
> +type ScaleSpec struct {
> +	// Replicas is the number of desired replicas.
> +	Replicas int
> +}
> +
> +// ScaleStatus represents the current status of a Scale subresource.
> +type ScaleStatus struct {
> +	// Replicas is the number of actual replicas.
> +	Replicas int
> +
> +	// Selector is a label query over pods that should match the replicas count.
> +	Selector map[string]string

@ramr <https://github.com/ramr> @davidopp <https://github.com/davidopp>
@bgrant0607 <https://github.com/bgrant0607>
We decided that instead of hard-coding all types that autoscaler can work
on (RC, deployment, other future types?) we will implement a generic way of
scaling different types of object: scale subresource. The wider discussion
regarding scale took place in #1629
<https://github.com/kubernetes/kubernetes/issues/1629>. The main advanatage
of scale resource is that whenever we introduce another type we want to
auto-scale, we just need to implement scale for it (w/o modifying
autoscaler code or API).

However, as you correctly noticed, scale subresource has some drawback
because some fields become redundant with what we already have in RC
object: Spec.Replicas, Status.Replicas and Status.Selector. In my opinion,
we should get rid of the redundancy in future by removing redundant fields
from RC.Spec/RC.Status. We should also introduce Deployment object in a way
that it doesn't have the redundant fields (but relies on Scale). Unlucky,
proposed changes in RC will break v1 API, so, proper converters will need
to be written and both versions should be supported for some time.


As a client, I should be able to atomically create an RC and set replicas
and selector.  The scale resource is a different view of the RC -
redundancy is not an issue.  If scale clients need that info, they should
be accessing the scale resource (interface) for it.  RC is the type, Scale
is an interface.

Or is the concern just duplicating the Go struct types?


If we all agree that removing the redundancy is the way to go, I'll update
the design doc and describe it. Please note that autoscaler API will be in
experimental for some time, and there are also other obvious changes we
need to introduce before moving it v1/v2 (e.g.: custom metrics).

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12859/files#r37546968>.